Let torch dynamo inline torch.func.grad

guilhermeleobas opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120338
* #119926
* #119405
* __->__ #118407

When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

guilhermeleobas committed:
Let torch dynamo inline torch.func.grad

When dynamo sees torch.func.grad, it tries to inline all frames related
to.

[ghstack-poisoned]

guilhermeleobas commented:
Still a work in progress.

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas commented:
@zou3519, PR is only missing a guard on `grad_increment_nesting` context manager, which I will add after fix https://github.com/pytorch/pytorch/issues/118493. Other than that, PR should be ready for review.

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented on test/functorch/test_eager_transforms.py:
```diff
@@ -4879,14 +4830,17 @@ def wrapper_fn(x, y):
         actual = wrapper_fn(x, y)
         expected = torch.compile(wrapper_fn, backend='eager', fullgraph=True)(x, y)
-        self.assertEqual(actual, expected)
+        fn = torch.compile(wrapper_fn, backend='eager', fullgraph=True)
+        import dis
+        print(dis.dis(fn))
+        # self.assertEqual(actual, expected)
```
What's going on here?

guilhermeleobas replied:
Ops, my bad! Forgot to remove this one.

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -723,7 +723,8 @@ def step(self):
                         # instruction is not in another block, then there
                         # should be no other nested blocks that we are in.
-                        assert len(self.block_stack) == 1
+                        # assert len(self.block_stack) == 1  # this is no longer true?
+                        # pytest -rs -sv test/functorch/test_eager_transforms.py -k test_grad_deprecated_api_cpu
```
cc @williamwen42 is this assertion safe to remove?

williamwen42 replied:
No - why are we removing this?

guilhermeleobas replied:
@williamwen42, In the test case below, the length of `block_stack` is 2.

```
import torch
x = torch.randn(())
y = torch.randn(())

@torch.compile(backend='eager', fullgraph=True)
def wrapper_fn(x, y):
    return torch.func.grad(torch.mul)(x, y)

wrapper_fn(x, y)
```

If I comment the assertion, code works just fine.

zou3519 replied:
@guilhermeleobas do you have a test case that does not involve grad?

williamwen42 replied:
We'll need to look more into why block stack is 2 in this case - I don't expect the assertion to fail.

guilhermeleobas replied:
@williamwen42, could you elaborate a little bit on why the block stack here is supposed to be 1? Is it related to a context manager?

williamwen42 replied:
Yes, the block stack is used to keep track of active context managers/try blocks/etc. There is one block for each nested context manager that we are in. The assert is there because if the current instruction is not in any block, we expect that we were previously only 1 context manager/block deep. A block stack size of 2 at this point means we're jumping out of 2 context managers at once, which is something I don't expect to happen.

guilhermeleobas replied:
Ok. Perhaps this explains the failure. There are two context managers in `grad`. See:
https://github.com/pytorch/pytorch/blob/ee4d3cac83345ebfecc60ba1a2c38f1a31df005a/torch/_functorch/eager_transforms.py#L1232-L1235

williamwen42 replied:
I don't expect it because context managers have cleanup bytecode after running, so we should never jump more than 1 context manager out at a time. Example (in particular, the triple `LOAD_CONST None` is context manager cleanup bytecode):
```python
import dis

def fn():
    with X():
        with Y():
            a = b

dis.dis(fn)
```
Output:
```
  3           0 RESUME                   0

  4           2 LOAD_GLOBAL              1 (NULL + X)
             14 PRECALL                  0
             18 CALL                     0
             28 BEFORE_WITH
             30 POP_TOP

  5          32 LOAD_GLOBAL              3 (NULL + Y)
             44 PRECALL                  0
             48 CALL                     0
             58 BEFORE_WITH
             60 POP_TOP

  6          62 LOAD_GLOBAL              4 (b)
             74 STORE_FAST               0 (a)

  5          76 LOAD_CONST               0 (None)
             78 LOAD_CONST               0 (None)
             80 LOAD_CONST               0 (None)
             82 PRECALL                  2
             86 CALL                     2
             96 POP_TOP
             98 JUMP_FORWARD            11 (to 122)
        >>  100 PUSH_EXC_INFO
            102 WITH_EXCEPT_START
            104 POP_JUMP_FORWARD_IF_TRUE     4 (to 114)
            106 RERAISE                  2
        >>  108 COPY                     3
            110 POP_EXCEPT
            112 RERAISE                  1
        >>  114 POP_TOP
            116 POP_EXCEPT
            118 POP_TOP
            120 POP_TOP

  4     >>  122 LOAD_CONST               0 (None)
            124 LOAD_CONST               0 (None)
            126 LOAD_CONST               0 (None)
            128 PRECALL                  2
            132 CALL                     2
            142 POP_TOP
            144 LOAD_CONST               0 (None)
            146 RETURN_VALUE
        >>  148 PUSH_EXC_INFO
            150 WITH_EXCEPT_START
            152 POP_JUMP_FORWARD_IF_TRUE     4 (to 162)
            154 RERAISE                  2
        >>  156 COPY                     3
            158 POP_EXCEPT
            160 RERAISE                  1
        >>  162 POP_TOP
            164 POP_EXCEPT
            166 POP_TOP
            168 POP_TOP
            170 LOAD_CONST               0 (None)
            172 RETURN_VALUE
ExceptionTable:
  30 to 58 -> 148 [1] lasti
  60 to 74 -> 100 [2] lasti
  76 to 98 -> 148 [1] lasti
  100 to 106 -> 108 [4] lasti
  108 to 112 -> 148 [1] lasti
  114 to 114 -> 108 [4] lasti
  116 to 120 -> 148 [1] lasti
  148 to 154 -> 156 [3] lasti
  162 to 162 -> 156 [3] lasti
```

williamwen42 replied:
Can you share what the bytecode + exception table looks like at the time of the assert fail?

guilhermeleobas replied:
Thank you. I just discovered that if I replace the `aux is not None` condition by `has_aux`, this issue gets fixed.
https://github.com/pytorch/pytorch/blob/5a0a964444d6be3b2762d30f5aec456060fa63c8/torch/_functorch/eager_transforms.py#L1287-L1290

I'll continue debugging this, but I guess this won't be a blocker anymore if CI passes.

guilhermeleobas replied:
> Can you share what the bytecode + exception table look like at the time of the assert fail?

Sure! here
https://gist.github.com/guilhermeleobas/8503cb67441b56fdf057ce120d31215e

williamwen42 replied:
Hmm, unsure why block_stack would be 2 here since the only instruction that would cause a block push is BEFORE_WITH and we don't jump backward to it.

guilhermeleobas replied:
Should I open a new issue to track this bug? CI is green with latest change and perhaps one can merge this one.

williamwen42 replied:
Sure. I'll defer the review back to @zou3519.

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -2098,18 +2099,18 @@ def __init__(
+            torch._C._functorch.TransformType.Vmap,
+            torch._C._functorch.TransformType.Grad,
+        )
+        if ci is not None and ci.key() in forbidden_keys and compiler_fn is not eager:
```
Need to update "Fallback to eager in case of a graph break inside vmap" comment

zou3519 replied:
And the unimplemented msg

zou3519 commented on torch/_subclasses/meta_utils.py:
```diff
@@ -327,14 +329,30 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
                             lvl = maybe_get_level(t)
                             bdim = maybe_get_bdim(t)
                             r = _add_batch_dim(ft, bdim, lvl)
+                        elif is_gradtrackingtensor(t):
```
"# Wraps a BatchedTensor in a FakeTensor" comment needs updating

zou3519 commented on torch/_subclasses/meta_utils.py:
```diff
@@ -327,14 +329,30 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
                             bdim = maybe_get_bdim(t)
                             r = _add_batch_dim(ft, bdim, lvl)
+                        elif is_gradtrackingtensor(t):
+                            r = self.meta_tensor(
+                                get_unwrapped(t),
+                                shape_env=shape_env,
+                                callback=callback,
+                                source=source,
+                                symbolic_context=symbolic_context,
+                            )
+                            assert is_gradtrackingtensor(r) is False, r
```
I'm confused, the result of this should be a GradTrackingTensor holding a FakeTensor

zou3519 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -378,13 +379,15 @@ def mk_fake_tensor(make_meta_t):
-            symbolic_context=symbolic_context,
-        )
+        disable_functorch = torch._C._DisableFuncTorch
+        with disable_functorch():
```
Hmm, this might be concerning (that we need to turn off functorch at random points in Dynamo). let me think about this some more.

guilhermeleobas replied:
If we don't turn off functorch and grad is enabled, any torch.* op in this file returns a GradTrackingTensor.

zou3519 replied:
Maybe let's try to fix the TensorWrapper fake-ification (that I mentioned in the other comments) and see if this is still necessary.

zou3519 commented on torch/_dynamo/variables/ctx_manager.py:
```diff
@@ -150,6 +150,75 @@ def exit(self, tx, *args):
 
 
+class GradInplaceRequiresGradCtxManagerVariable(ContextWrappingVariable):
+    """represents torch grad requries grad"""
```
nit: more description?

zou3519 reviewed (commented):
This mostly LGTM. Things we should take a look at:
- fakeification of the TensorWrapper (GradTrackingTensor)
- if we can do something more systematic about needing to put disable_functorch in various places
- guards for the functorch stack. Now that we support vmap and grad, we should try to see if our guards work in all situations. E.g. someone may call `grad` inside of a vmap + grad -- we should probably guard on the state of the entire functorch stack.

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

izaitsevfb commented:
@pytorchbot drci

(want to check whether mergeability check is marked as stable after recent config change)

guilhermeleobas commented:
@zou3519, if a GradTrackingTensor `t` is a view, its fake version should also be a view?

```python
>>> t
GradTrackingTensor(lvl=2, value=
    tensor([ 0.0461,  0.4024, -1.0115])
)
>>> t._is_view()
True
>>> r = self.meta_tensor(t, shape_env, callback, source, symbolic_context)
>>> r
GradTrackingTensor(lvl=2, value=
    FakeTensor(..., size=(3,))
)
>>> r._is_view()
# Ought this to be True?
```

zou3519 commented:
Yes

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas commented:
Current lint failure is not related to this PR

zou3519 commented on torch/_subclasses/meta_utils.py:
```diff
@@ -327,16 +329,39 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
                             r = _add_batch_dim(ft, bdim, lvl)
+                        elif is_gradtrackingtensor(t):
+                            disable_functorch = torch._C._DisableFuncTorch
+                            with disable_functorch():
+                                ft = _to_fake_tensor(get_unwrapped(t))
+                            lvl = torch._C._functorch.maybe_get_level(t)
+                            r = torch._C._functorch._wrap_for_grad(ft, lvl)
+
+                            if t._is_view():
+                                # is this correct?
+                                # it fails if r is a nested functorch tensor:
+                                # r = GradTrackingTensor(
+                                #       BatchedTensor(
+                                #           Tensor
+                                #       )
+                                #   )
+                                r = r.view_as(r)
```
From my understanding, FakeTensor has a mechanism for modeling view relationships. Could we figure out how it does it for regular tensors and figure out how to apply it to GradTrackingTensor?

zou3519 replied:
Hmm, I have a better question: when do we actually run this fakeification logic for GradTrackingTensor?

I would expect that FakeTensor propagation (e.g. calling GradTrackingTensor(FakeTensor).sin() just works). So it's unclear to me how we get to this point

zou3519 commented on torch/_dynamo/variables/ctx_manager.py:
```diff
@@ -150,6 +150,87 @@ def exit(self, tx, *args):
+
+    def enter(self, tx):
+        grad_level = torch._C._functorch._grad_increment_nesting()
+        install_guard(self._guards_singleton)
```
Why do we install the guard after we increment the level?

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -3064,6 +3306,7 @@ def fn(x):
         # should not recompile on second call. See Pytorch issue #118493
         y = torch.vmap(fn)(x)
 
+    @xfailIfTorchDynamo
```
What's going on here?

zou3519 reviewed (commented):
Looks mostly good to me. A couple of things:
- the fakeification looks suspicious, we should go understand how regular Fakeification with views looks
- there are a number of xfails added to tests involving recompiles. Those don't look OK. Were we planning on investigating them?

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas commented:
> the fakeification looks suspicious, we should go understand how regular Fakeification with views looks

I've changed the fakeification of GradTrackingTensor to raise `unimplemented` on the cases the input tensor is a view.

> there are a number of xfails added to tests involving recompiles. Those don't look OK. Were we planning on investigating them?

I am investigating them but what happen is that the exception is not raised when `PYTORCH_TEST_WITH_DYNAMO=1`

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented:
This PR checked in some submodules, we need to remove them

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -2377,6 +2375,60 @@ def fn(pred, pytree_in):
+    - torch._functorch.pyfunctorch.retrieve_current_functorch_interpreter().check_state((1,))  # with grad_increment_nesting() as level:  # _functorch/eager_transforms.py:1232 in grad_and_value_impl""",
+            record.getMessage(),
+        )
+
```
Can we add a test that calling grad(fn)(x) again does not recompile?

zou3519 reviewed (commented):
I think the only blocker is resolving with @williamwen42 the removal of the assertion. Left some other comments around testing.
The new xfails seem reasonable to take as a follow-up (could we do a brief sanity check that the error messages do not indicate silent incorrectness?)

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -2377,6 +2375,60 @@ def fn(pred, pytree_in):
+        fn(x)
+        self.assertEqual(len(records), 0)
+
+    @unittest.expectedFailure
```
probably need to remove this

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas commented:
@pytorchbot merge

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Let torch dynamo inline torch.func.grad"


When dynamo sees torch.func.grad, it tries to inline all frames related
to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]
